src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Summarize.cs

```diff
@@ -0,0 +1,245 @@
+            ScopeInfo = new FunctionThisGroupScopeInfo(this);
+        }
+
+        public override IEnumerable<TexlStrings.StringGetter[]> GetSignatures()
+        {
+            yield return new[] { TexlStrings.SummarizeArg1, TexlStrings.SummarizeArg2, TexlStrings.SummarizeArg3 };
+            yield return new[] { TexlStrings.SummarizeArg1, TexlStrings.SummarizeArg2, TexlStrings.SummarizeArg3, TexlStrings.SummarizeArg2 };
+            yield return new[] { TexlStrings.SummarizeArg1, TexlStrings.SummarizeArg2, TexlStrings.SummarizeArg2, TexlStrings.SummarizeArg3, TexlStrings.SummarizeArg3 };
+        }
+
+        // Produce a signature of the form:
+        // Sumamrize(source, grouping_column, aggregation, grouping_column, aggregation, ...)
+        public override IEnumerable<TexlStrings.StringGetter[]> GetSignatures(int arity)
+        {
+            Contracts.Assert(arity >= 0);
+
+            if (arity <= 3)
+            {
+                return base.GetSignatures(arity);
+            }
+
+            // Limit the argCount avoiding potential OOM
+            int argCount = arity > SignatureConstraint.RepeatTopLength + SignatureConstraint.EndNonRepeatCount ? SignatureConstraint.RepeatTopLength + SignatureConstraint.EndNonRepeatCount : arity;
+            var args = new TexlStrings.StringGetter[argCount];
+
+            int lastIndex = argCount - 1;
+            args[0] = TexlStrings.SummarizeArg1;
+
+            var addSwitch = false;
+
+            for (int i = 1; i < lastIndex; i++)
+            {
+                if (addSwitch)
+                {
+                    args[i] = TexlStrings.SummarizeArg2;
+                }
+                else
+                {
+                    args[i] = TexlStrings.SummarizeArg3;
+                }
+
+                addSwitch = !addSwitch;
+            }
+
+            return new[] { args };
+        }
+
+        public override bool CheckTypes(CheckTypesContext context, TexlNode[] args, DType[] argTypes, IErrorContainer errors, out DType returnType, out Dictionary<TexlNode, DType> nodeToCoercedTypeMap)
+        {
+            Contracts.AssertValue(args);
+            Contracts.AssertAllValues(args);
+            Contracts.AssertValue(argTypes);
+            Contracts.Assert(args.Length == argTypes.Length);
+            Contracts.AssertValue(errors);
+            Contracts.Assert(MinArity <= args.Length && args.Length <= MaxArity);
+
+            bool isValid = base.CheckTypes(context, args, argTypes, errors, out returnType, out nodeToCoercedTypeMap);
+            Contracts.Assert(returnType.IsTable);
+
+            // Source table type
+            DType sourceType = argTypes[0];
+
+            if (args[0] is AsNode)
+            {
+                isValid = false;
+                errors.EnsureError(DocumentErrorSeverity.Severe, args[0], TexlStrings.ErrSummarizeDataSourceScopeNotSupported);
+            }
+
+            if (sourceType.Contains(FunctionThisGroupScopeInfo.ThisGroup))
```

Comment:
>if (sourceType.Contains(FunctionThisGroupScopeInfo.ThisGroup)) [](http://example.com/codeflow?start=12&length=62)

Do we also need to check if any of the display names is named 'ThisGroup'? In the test SummarizeDisplayNamesTest we check that we cannot use a column whose display name is ThisGroup as a grouping column. But should it work if we didn't use it?